sound/soc/sof/topology.c

```diff
@@ -2795,3 +2795,16 @@ int snd_sof_load_topology(struct snd_sof_dev *sdev, const char *file)
 }
 EXPORT_SYMBOL(snd_sof_load_topology);
 
+void snd_sof_free_topology(struct snd_sof_dev *sdev)
+{
+	int ret;
+
+	dev_dbg(sdev->dev, "free topology...\n");
+
+	ret = snd_soc_tplg_component_remove(sdev->component,
+					    SND_SOC_TPLG_INDEX_ALL);
+	if (ret < 0)
+		dev_err(sdev->dev,
+			"error: tplg component free failed %d\n", ret);
+}
+EXPORT_SYMBOL(snd_sof_free_topology);
```

Comment:
sorry @ranj063 you convinced me everything was handled by the framework and now we add this back? Why? the commit message "This is important to free the topology resources before the card is freed when the machine driver is unregistered." makes no sense to me.

Reply:
@plbossart I know this is controversial. But let me explain my rationale:

1. If we manually unregister the component driver as we do today in SOF, this is PR is not relevant at all. We make sure the component is unregistered before the machine driver is unregistered. And unregistering the machine driver automatically frees topology, which is why we removed it last week.

But, if we want to move towards using devm_snd_soc_register_component() ( which I think makes sense and somebody might suggest to make the change when we upstream), we will run into issues with the card being freed before the topology is freed.

1. with devm_snd_soc_regsiter_component(), the unregister will happen after sof_remove().
2. Since, sof_remove()  unregisters the machine driver first, the card will be freed first when dealing with the topology component removal, we'll run into null pointer exception since the card has been freed.
3. So, if we want to move to devm_snd_soc_register_component(), we need to make sure we free the topology before we unregister the machine driver.